File: src/main/python/dfsbfs/보물섬2589.py
import sys
from collections import deque

input = sys.stdin.readline
n, m = map(int, input().split())
graph = [list(input().rstrip()) for _ in range(n)]

# 육지 개수 확인 - 1개 이하면 바로 종료
land_count = sum(row.count('L') for row in graph)
if land_count <= 1:
  print(0)
  exit()

# 방향 배열 미리 선언
dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

def bfs(start_y, start_x):
  visited = set([(start_y, start_x)])
  dist = [[0] * m for _ in range(n)]
  q = deque([(start_y, start_x)])
  max_dist = 0

  while q:
    ey, ex = q.popleft()
    for i in range(4):
      ny, nx = ey + dy[i], ex + dx[i]
      if 0 <= ny < n and 0 <= nx < m and graph[ny][nx] == 'L' and (ny, nx) not in visited:
        visited.add((ny, nx))
        dist[ny][nx] = dist[ey][ex] + 1
        max_dist = max(max_dist, dist[ny][nx])
        q.append((ny, nx))

  return max_dist

maxi = 0
for y in range(n):
  for x in range(m):
    if graph[y][x] == 'L':
      maxi = max(maxi, bfs(y, x))

print(maxi)

File: src/main/java/org/snust/dfsbfs/보물섬2589.java
package org.snust.dfsbfs;

import java.io.*;
import java.util.*;

public class 보물섬2589 {
  static int n, m;
  static char[][] graph;
  static int[] dx = {0, 0, 1, -1};
  static int[] dy = {1, -1, 0, 0};

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());

    n = Integer.parseInt(st.nextToken());
    m = Integer.parseInt(st.nextToken());
    graph = new char[n][m];

    // 그래프 입력
    int landCount = 0;
    for(int i = 0; i < n; i++) {
      String line = br.readLine();
      for(int j = 0; j < m; j++) {
        graph[i][j] = line.charAt(j);
        if(graph[i][j] == 'L') landCount++;
      }
    }

    // 육지가 1개 이하면 종료
    if(landCount <= 1) {
      System.out.println(0);
      return;
    }

    int maxi = 0;
    for(int y = 0; y < n; y++) {
      for(int x = 0; x < m; x++) {
        if(graph[y][x] == 'L') {
          maxi = Math.max(maxi, bfs(y, x));
        }
      }
    }

    System.out.println(maxi);
  }

  static int bfs(int startY, int startX) {
    Set<Point> visited = new HashSet<>();
    int[][] dist = new int[n][m];
    Queue<Point> q = new LinkedList<>();

    Point start = new Point(startY, startX);
    visited.add(start);
    q.offer(start);
    int maxDist = 0;

    while(!q.isEmpty()) {
      Point cur = q.poll();
      for(int i = 0; i < 4; i++) {
        int ny = cur.y + dy[i];
        int nx = cur.x + dx[i];
        Point next = new Point(ny, nx);

        if(ny >= 0 && ny < n && nx >= 0 && nx < m
                && graph[ny][nx] == 'L' && !visited.contains(next)) {
          visited.add(next);
          dist[ny][nx] = dist[cur.y][cur.x] + 1;
          maxDist = Math.max(maxDist, dist[ny][nx]);
          q.offer(next);
        }
      }
    }
    return maxDist;
  }

  static class Point {
    int y, x;
    Point(int y, int x) {
      this.y = y;
      this.x = x;
    }

    @Override
    public boolean equals(Object o) {
      if(this == o) return true;
      if(!(o instanceof Point)) return false;
      Point point = (Point) o;
      return y == point.y && x == point.x;
    }

    @Override
    public int hashCode() {
      return Objects.hash(y, x);
    }
  }
}
